wallet/wallet.go

```diff
@@ -509,43 +523,71 @@ func (w *Wallet) SetAgendaChoices(ctx context.Context, ticketHash *chainhash.Has
 }
 
 // TreasuryKeyPolicy returns a vote policy for provided Pi key. If there is
-// no policy return TreasuryVoteInvalid.
-func (w *Wallet) TreasuryKeyPolicy(pikey []byte) stake.TreasuryVoteT {
+// no policy this method returns TreasuryVoteInvalid.
+// A non-nil ticket hash may be used by a VSP to return per-ticket policies.
+func (w *Wallet) TreasuryKeyPolicy(pikey []byte, ticket *chainhash.Hash) stake.TreasuryVoteT {
 	w.stakeSettingsLock.Lock()
 	defer w.stakeSettingsLock.Unlock()
 
-	// Zero value means abstain, just return as is.
+	// Zero value is abstain/invalid, just return as is.
+	if ticket != nil {
+		return w.vspTSpendKeyPolicy[udb.VSPTreasuryKey{
+			Ticket:      *ticket,
+			TreasuryKey: string(pikey),
+		}]
+	}
 	return w.tspendKeyPolicy[string(pikey)]
 }
 
 // TSpendPolicy returns a vote policy for a tspend.  If a policy is set for a
 // particular tspend transaction, that policy is returned.  Otherwise, if the
 // tspend is known, any policy for the treasury key which signs the tspend is
 // returned.
-func (w *Wallet) TSpendPolicy(hash *chainhash.Hash) stake.TreasuryVoteT {
+// A non-nil ticket hash may be used by a VSP to return per-ticket policies.
+func (w *Wallet) TSpendPolicy(tspendHash, ticketHash *chainhash.Hash) stake.TreasuryVoteT {
 	w.stakeSettingsLock.Lock()
 	defer w.stakeSettingsLock.Unlock()
 
-	if policy, ok := w.tspendPolicy[*hash]; ok {
+	// Policy preferences for specific tspends override key policies.
+	if ticketHash != nil {
+		policy, ok := w.vspTSpendPolicy[udb.VSPTSpend{
+			Ticket: *ticketHash,
+			TSpend: *tspendHash,
+		}]
+		if ok {
+			return policy
+		}
+	} else if policy, ok := w.tspendPolicy[*tspendHash]; ok {
```

Comment:
This can't be an `else if`, it needs to be a separate `if`, otherwise if you set a global policy for a specific tspend (that is, without specifying a ticket hash) the wallet won't vote for it because the `ticketHash` argument is always filled now.